R_notebooks/03_Manipulating_a_dataframe.Rmd

```diff
@@ -0,0 +1,431 @@
+
+For example, to get the average flipper length of each sex:
+
+```{r}
+# First group the data by sex
+penguins_grouped_by_sex <- dplyr::group_by(penguins, sex)
+
+# Then find the mean flipper length for each sex
+dplyr::summarise(penguins_grouped_by_sex, mean_flipper_length_mm = mean(flipper_length_mm, na.rm = TRUE))
+```
+
+
+These functions are powerful. We can group by multiple columns at once to create
+pairwise groups. As we saw above, we can also create several summary variables
+at the same time:
+
+```
+dplyr::group_by(data, col1, col2, ...)  # group by combinations of values in
+                                        # col1, col2, ...
+
+# Compute multiple summaries on the same data
+dplyr::summarise(grouped_data, summary1 = <...>, summary2 = <...>, ...)
+```
+
+
+### Exercise: grouping and summarising
+
+Create a summary dataframe that gives the mean, range (i.e. `max - min`)
+and standard deviation of the body mass for each species / sex combination, as
+well as the number of observations in each group.
+
+```{r}
+penguins_grouped <- dplyr::group_by(penguins, species, sex)
+dplyr::summarise(penguins_grouped,
+                 mean_body_mass_g = mean(body_mass_g, na.rm = TRUE),
+                 sd_body_mass_g = sd(body_mass_g, na.rm = TRUE),
+                 range_body_mass_g = max(body_mass_g, na.rm = TRUE) - min(body_mass_g, na.rm = TRUE))
+```
+
+
+### An unimportant detail: grouped dataframes
+
+`dplyr::group_by` returns a 'grouped dataframe' (`dplyr::grouped_df`). These
+behave just like tibbles/dataframes except they have some extra information
+about the grouping attached to them.
+
+When you group by multiple columns, the result of `summarise` will typically
+also be a grouped dataframe. Usually, the groups are given by all columns except
+the last one in the grouped dataframe supplied to `summarise`. By default, you
+get a console message telling you on which columns the output is grouped by (see
+the result of the exercise above).
+
+The upshot of this is that it allows us to do successive summaries without
+having to keep regrouping the data. For example, the code below will produce a
+'max of averages':
+
+```{r}
+penguins_grouped <- dplyr::group_by(penguins, sex, species)  # grouped by sex and species
+
+# Average mass for each species / sex combination
+# Result is grouped by sex (see console output)
+penguins_grouped_mass <- dplyr::summarise(penguins_grouped,
+                                          mean_body_mass_g = mean(body_mass_g, na.rm = TRUE))
+print(penguins_grouped_mass)
+
+# Max of the average mass for each species
+penguins_max_species_avg_mass <- dplyr::summarise(
+  penguins_grouped_mass,
+  max_species_avg_body_mass_g = max(mean_body_mass_g, na.rm = TRUE)
+  )
+print(penguins_max_species_avg_mass)
+```
+
+In practice, you don't have to worry about this detail. If you want to get back
+to a plain old tibble/dataframe, use the `dplyr::ungroup` function.
```

Comment:
I wonder if it is worth being explicit here that ungroup only works on grouped data not a datatable that has had summarise applied to it (where it is my understanding that a datatable that has been modified by summarise reduces its dimensions and therefore can't be ungrouped?).

Reply:
Thanks for pointing that out. I've checked and it turns out that `summarise` can in fact create a grouped dataframe, if grouping by more than one variable. I've made the clarification about applying `ungroup` to a grouped dataframe.